### ISSUE TITLE
npm i deletes all the symlinked repos

### ISSUE THREAD

**Issue by yawhide** [npm/npm#17257]
#### I'm opening this issue because:

  - [ ] npm is crashing.
  - [x] npm is producing an incorrect install.
  - [x] npm is doing something I don't understand.
  - [ ] Other (_see below for feature requests_):

#### What's going wrong?
When I do npm link B inside repo A, npm deletes all the node_modules inside B.

#### How can the CLI team reproduce the problem?
- git clone repo A and B
- call `npm i` inside repo A
- call `npm i` inside repo B
- repo A has a dependency on repo B (but nothing is symlinked yet)
- call `npm link` in repo B.
- go to repo A, call `npm link B`.
- call `npm i` in repo A. <---- this deletes a/node_modules AND b/node_modules but it shouldn't delete b/node_modules.
- go back to repo B and observe that the node-modules folder is empty.
- the symlink to repo B from repo A is still preserved (thankfully) so all you have to do is go back to repo B and call `npm i`.

this is especially annoying if I add a dependency in repo A, call `npm i lodash` for example, which then deletes all the symlinked repos...

### supporting information:

 - `npm -v` prints: 5.0.3
 - `node -v` prints: v8.1.2
 - `npm config get registry` prints: https://registry.npmjs.org/
 - Windows, OS X/macOS, or Linux?: Linux/Ubuntu
 - Network issues:
   - Geographic location where npm was run:
   - [ ] I use a proxy to connect to the npm registry.
   - [ ] I use a proxy to connect to the web.
   - [ ] I use a proxy when downloading Git repos.
   - [ ] I access the npm registry via a VPN
   - [ ] I don't use a proxy, but have limited or unreliable internet access.
 - Container:
   - [ ] I develop using Vagrant on Windows.
   - [ ] I develop using Vagrant on OS X or Linux.
   - [ ] I develop / deploy using Docker.
   - [ ] I deploy to a PaaS (Triton, Heroku).

**Comment by yawhide** [comment]
basically if you have a symlink in node_modules and you call npm i, npm i deletes the current node_modules AND goes into the symlink'ed node_modules and deletes it. VERY annoying.

**Comment by onury** [comment]
Does this happen on older versions? I'm tired of re-installing removed packages and re-linking..

**Comment by yawhide** [comment]
i cant repro this in npm v4.2.0

**Comment by ile** [comment]
It seems that npm 5.0.3 just deleted all the git repositories in node_modules, symlinked or not. I did `npm i some-module`.

Not good :(

Edit: yes it did, and I lost some code. Was able to retrieve some of it from editor's undo history. :grimacing:

**Comment by ghost** [comment]
I just ran into this issue after upgrading. I'm on npm 5.0.3.

When I make a symlink to some other module in my ~/projects directory then I use `npm install` to install another package, the symlink disappears and the node_modules inside my project directory disappears too:

```
$ ln -s ~/projects/hyperdb node_modules
$ ls -d node_modules/hyperdb
node_modules/hyperdb@
$ ls -d ~/projects/hyperdb/node_modules
/home/substack/projects/hyperdb/node_modules/
$ npm i random-access-memory
+ random-access-memory@2.4.0
removed 1 package and updated 1 package in 3.337s
$ ls -d node_modules/hyperdb
ls: cannot access node_modules/hyperdb: No such file or directory
$ ls -d ~/projects/hyperdb/node_modules
ls: cannot access /home/substack/projects/hyperdb/node_modules: No such file or directory
```

**Comment by mreinstein** [comment]
I ran into the same issue as @substack. Also on npm 5.0.3. This is a somewhat common usage pattern in node/browserify, and it's referenced in the handbook. https://github.com/substack/browserify-handbook#avoiding-

**Comment by iarna** [comment]
As of `npm@5.1.0` the behavior is now this:

It replaces your symlink with a registry version (as symlinks are a thing you can ask for in `package.json`'s now) but it doesn't delete the linked project's dependencies, which means you can link again without quite so much hair pulling.

To properly and completely fix this, I believe we'll need changes we've been discussing to `npm link` that will make note of development links like this so it knows to maintain them.

Basically, it thinks you previously had a `file:` spec in your `package.json` and switched to a version range and so it's updating your `node_modules` to use the version range. Fixing this means being able to distinguish between that scenario and the one described in this issue.

**Comment by ghost** [comment]
Does the planned fix mean I will have to start using `npm link` instead of `ln -s`? I have so far avoided using `npm link` because it does a lot of other things besides creating a symlink in node_modules. Some of those things seem to involve a network round-trip or recompiling deps.

**Comment by iarna** [comment]
The plan is to make `npm link` work largely like `ln -s`. That is, `npm link ../abc-def` would create a symlink to `../abc-def` in your `node_modules` and record this fact, but wouldn't do global installs or anything else. Before we move forward with that we'll have an RFC up.

**Comment by mikermcneil** [comment]
Seems like the package-lock file will prevent some of the roundtripping. But I've definitely seen some other unexpected deletions of node_modules folders when doing npm link lately (I've told folks on my team to just delete their package-lock file, delete node_modules and then `npm install` fresh -- but that doesn't help for the folks that need to use npm link)

Re: the symlink deletion, I've noticed the same thing as @substack since npm 3.  In the Sails CLI, we ended up giving up and just forcing a new npm install.  But we used to do things this way, so that generating a new project didn't mean reinstalling everything again: https://github.com/balderdashy/sails-generate/blob/4ac5c14636b5ed18e4b144c0df8d4a33a9ccba14/lib/core-generators/new/after.js#L153-L263

**Comment by avesus** [comment]
> Before we move forward with that we'll have an RFC up.

@iarna how's the RFC doing?.. I develop a library and an example app depending on the library. And I have auto-deploy of the example app on Heroku each time I commit it on GitHub.

I have two issues with *new* NPM:
1) package-lock.json fixes my library's commit hash, and I need to update it each time I commit the library. Sometimes I forget to do that and debug weird bugs which weren't happen with my workflow in NPM 4.
2) each time after installing a new dependency at the example app, I have to type npm link <my_local_library_package> which is annoying, but I can live with that because when debugging locally, it just crashes; on Heroku it uses real reference (but see (1) - it uses old commit hash).

Removing the library reference from package-lock.json of the example app expected to help to solve (1) but it actually doesn't help. Each time I install or remove a package from the example app, NPM 5.0.3 restores the entry with exact commit hash at package-lock.json, and my commit of the example app have to be happened strictly after commiting of the updated version of my library. And each time I'm doing a commit of the example app, I have to type npm install <my_library> after commiting the library on GitHub.

Why so many complications for this very *basic* process of libraries development just after introducing mechanisms to support integrity (package-lock.json) and `file:` dependencies?

It's disappointing that developers should spend so much time just because you guys, when introducing these _cool_ features (I really love both), forget to think about feature interactions with the very core behavior that even @substack was concerned!

**Comment by mreinstein** [comment]
@avesus 

>  you guys...forget to think about feature interactions with the very core behavior 
> that even @substack was concerned!

keep in mind npm has been around for a while, and has many, many different feature interactions. I know this wasn't intentionally overlooked. I don't see a way around this, it's just the nature of complexity.

**Comment by mreinstein** [comment]
@iarna curious, has there been any movement on this?

**Comment by onury** [comment]
This should be labeled **big-bug**!

**Comment by ghost** [comment]
I've created a work-around for this bug in this branch: https://github.com/substack/npm/tree/no-delete-symlinks

To use it, clone and `npm link` the repository and then do either `npm config set delete-symlinks false` or pass in `--no-delete-symlinks` when you `npm install`.

Related no-auto-prune patch: https://github.com/npm/npm/issues/16853#issuecomment-338450203

**Comment by fengmk2** [comment]
I think this big-bug cause by https://github.com/npm/npm/pull/17733

**Comment by ibc** [comment]
Hi NPM team, do you plan to fix this issue?

**Comment by legodude17** [comment]
@ibc They do, see npm/rfcs#3
